services/gateway_manager/src/main/java/com/futurewei/alcor/gatewaymanager/service/GatewayService.java

```diff
@@ -0,0 +1,187 @@
+import com.futurewei.alcor.gatewaymanager.dao.GWAttachmentRepository;
+import com.futurewei.alcor.gatewaymanager.dao.GatewayRepository;
+import com.futurewei.alcor.gatewaymanager.entity.GWAttachment;
+import com.futurewei.alcor.gatewaymanager.entity.ResourceType;
+import com.futurewei.alcor.gatewaymanager.entity.StatusEnum;
+import com.futurewei.alcor.web.entity.gateway.*;
+import com.futurewei.alcor.web.restclient.GatewayManagerRestClinet;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.context.annotation.ComponentScan;
+import org.springframework.stereotype.Service;
+
+import java.util.*;
+import java.util.function.Supplier;
+
+@Slf4j
+@Service
+@ComponentScan("com.futurewei.alcor.web.restclient")
+public class GatewayService {
+
+    @Autowired
+    private GatewayRepository gatewayRepository;
+
+    @Autowired
+    private GWAttachmentRepository gwAttachmentRepository;
+
+    @Autowired
+    private GatewayManagerRestClinet restClinet;
+
+    private static final String GATEWAY_DESCRIPTION = "internal gateway";
+    private static final String GATEWAY_SUFFIX_NAME = " gateway";
+    private static final String ATTACHMENT_NAME_PREFIX = "attachment - ";
+
+
+    public GatewayService() {
+    }
+
+    public GatewayInfo createGatewayInfo(String projectId, VpcInfo vpcInfo) throws Exception {
+
+        //TODO check whether the project type is zeta by projectId
+
+        // construct the GatewayEntity and Attachment
+        GatewayInfo gatewayInfo = new GatewayInfo();
+        GatewayEntity gatewayEntity = new GatewayEntity();
+        createGatewayAndAttachment(projectId, vpcInfo, gatewayInfo, gatewayEntity);
+
+        AsyncExecutor executor = new AsyncExecutor();
+        executor.runAsync(restClinet::createDPMCacheGateway, projectId, gatewayInfo);
+        executor.runAsync(restClinet::createVPCInZetaGateway, new VpcInfoSub(vpcInfo.getVpcId(), vpcInfo.getVpcVni()));
+        executor.runAsync((Supplier<Void>) () -> {
+            // wait result of all async
+            List<Object> result = executor.joinAllAsync();
+            // whether to rollback by checking result's size is 2
+            try {
+                updateDPMCacheGateway(result, gatewayInfo, gatewayEntity, projectId);
+            } catch (Exception e) {
+                log.info("update GatewayEntity status to READY failed, detail message: {}", e.getMessage());
+                // rollback GM's DB and ask Zeta to remove the corresponding gateway resources
+            }
+
+            try {
+                rollback(result, gatewayInfo, gatewayEntity, projectId);
+            } catch (Exception e) {
+                log.info("rollback failed, detail message: {}", e.getMessage());
+                // If rollback failed, we should raise an alarm or error.
+            }
+            return null;
+        });
+        return gatewayInfo;
+    }
```

Comment:
We can leave rollback to another PR and figure out a better generic way for rollback.